File: data_structures/02_BT_traversal/bt_traversal.py
from typing import Union, List, Deque
from collections import deque


class TreeNode:
    def __init__(self, val: int=0, left: Union['TreeNode', None]=None, right: Union['TreeNode', None]=None) -> None:
        self.val = val
        self.left = left
        self.right = right

    """ PROPERTIES """
    @property  # var = TreeNode.val
    def val(self) -> int:
        return self._val

    @val.setter  # TreeNode.val = new_val
    def val(self, new_val: int) -> None:
        if (new_val < -100) or (new_val > 100):
            raise ValueError("Constraints for 'value' violated")
        self._val = new_val

    @property  # var = TreeNode.left
    def left(self) -> Union['TreeNode', None]:
        return self._left

    @left.setter  # TreeNode.left = new_left
    def left(self, new_left: Union['TreeNode', None]) -> None:
        if isinstance(new_left, TreeNode) or new_left is None:
            self._left = new_left
        else:
            raise ValueError(f"Unsupported type {type(new_left)} for TreeNode.left")

    @property  # var = TreeNode.right
    def right(self) -> Union['TreeNode', None]:
        return self._right

    @right.setter  # TreeNode.right = new_right
    def right(self, new_right: Union['TreeNode', None]) -> None:
        if isinstance(new_right, TreeNode) or new_right is None:
            self._right = new_right
        else:
            raise ValueError(f"Unsupported type {type(new_right)} for TreeNode.right")


class Solution:
    def insertLevelOrder(self, arr: List[int]) -> Union[TreeNode, None]:
        if len(arr) == 0 or arr[0] is None: return None  # Handle empty or root = None case
        root = TreeNode(arr[0])
        q = deque([root])
        i = 1
        while q and i < len(arr):
            current = q.popleft()
            # Insert the left/right child if it is not None
            if i < len(arr) and arr[i] is not None:
                current.left = TreeNode(arr[i])
                q.append(current.left)
            i += 1  # Move to the next element in the array
            if i < len(arr) and arr[i] is not None:
                current.right = TreeNode(arr[i])
                q.append(current.right)
            i += 1  # Move to the next element in the array
        return root

    def preorderTraversal(self, root: Union[TreeNode, None]=None) -> List[int]:
        result = []
        self.recursion(root, result, "preorder")
        return result

    def inorderTraversal(self, root: Union[TreeNode, None]=None) -> List[int]:
        result = []
        self.recursion(root, result, "inorder")
        return result
    
    def postorderTraversal(self, root: Union[TreeNode, None]=None) -> List[int]:
        result = []
        self.recursion(root, result, "postorder")
        return result
    
    def levelorderTraversal(self, root: Union[TreeNode, None]=None) -> List[List[int]]:
        result = []
        if root is None: return result
        q = deque([root])
        while q:
            levelSize = len(q)
            currentLevel = []
            for i in range(levelSize):
                node = q.popleft()
                currentLevel.append(node.val)
                if node.left is not None: q.append(node.left)
                if node.right is not None: q.append(node.right)
            result.append(currentLevel)
        return result
    
    def recursion(self, node: Union[TreeNode, None], result: List[int], type: str) -> None:
        if not node: return
        if type == "preorder":
            result.append(node.val)
            self.recursion(node.left, result, "preorder")
            self.recursion(node.right, result, "preorder")
        elif type == "inorder":
            self.recursion(node.left, result, "inorder")
            result.append(node.val)
            self.recursion(node.right, result, "inorder")
        elif type == "postorder":
            self.recursion(node.left, result, "postorder")
            self.recursion(node.right, result, "postorder")
            result.append(node.val)

File: data_structures/02_BT_traversal/bt_traversal.cpp
#include <iostream>
#include <stdexcept>
#include <cmath>
#include <vector>
#include <string>
#include <queue>

using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : left(nullptr), right(nullptr) {
        setVal(x);
    }
    TreeNode(int x, TreeNode *left, TreeNode *right) : left(left), right(right) {
        setVal(x);
    }
private:
    void setVal(int v) {
        if (v < -1000 || v > 1000) throw std::runtime_error("Constraints for 'value' violated");
        val = v;
    }
};


class Solution {
public:
    // Function to insert nodes level-wise from an array in level-order format
    TreeNode* insertLevelOrder(const vector<int>& arr) {
        if (arr.empty() || arr[0] == -1000) return nullptr;  // Handle empty or root = nullptr case
        TreeNode* root = new TreeNode(arr[0]);
        queue<TreeNode*> q;
        q.push(root);
        int i = 1;
        while (!q.empty() && i < arr.size()) {
            TreeNode* current = q.front();
            q.pop();
            // Insert the left/right child if it is not -1000
            if (i < arr.size() && arr[i] != -1000) {
                current->left = new TreeNode(arr[i]);
                q.push(current->left);
            }
            i++;  // Move to the next element in the array
            if (i < arr.size() && arr[i] != -1000) {
                current->right = new TreeNode(arr[i]);
                q.push(current->right);
            }
            i++;  // Move to the next element in the array
        }
        return root;
    }
    vector<int> preorderTraversal(TreeNode *root) {
        vector<int> result;
        recursion(root, result, "preorder");
        return result;
    }
    vector<int> inorderTraversal(TreeNode *root) {
        vector<int> result;
        recursion(root, result, "inorder");
        return result;
    }
    vector<int> postorderTraversal(TreeNode *root) {
        vector<int> result;
        recursion(root, result, "postorder");
        return result;
    }
    vector<vector<int>> levelorderTraversal(TreeNode *root) {
        vector<vector<int>> result;
        if (!root) return result;
        queue<TreeNode*> q;
        q.push(root);
        while (!q.empty()) {
            int levelSize = q.size();
            vector<int> currentLevel;
            for (int i = 0; i < levelSize; ++i) {
                TreeNode* node = q.front();
                q.pop();
                currentLevel.push_back(node->val);
                if (node->left) q.push(node->left);
                if (node->right) q.push(node->right);
            }
            result.push_back(currentLevel);
        }
        return result;
    }
private:
    void recursion(TreeNode *node, vector<int>& result, string type) {
        if (!node) return;
        if (type == "preorder") {
            result.push_back(node->val);
            recursion(node->left, result, type);
            recursion(node->right, result, type);
        } else if (type == "inorder") {
            recursion(node->left, result, type);
            result.push_back(node->val);
            recursion(node->right, result, type);
        } else if (type == "postorder") {
            recursion(node->left, result, type);
            recursion(node->right, result, type);
            result.push_back(node->val);
        }
    }
};